controller/lib/hal/adc.cpp

```diff
@@ -0,0 +1,240 @@
+  }
+
+  // Clear all the status bits
+  adc->adc[0].stat = 0x3FF;
+
+  // Enable the A/D
+  adc->adc[0].ctrl |= 0x00000001;
+
+  // Wait for the ADRDY status bit to be set
+  while (!(adc->adc[0].stat & 0x00000001)) {
+  }
+
+  adc->adc[0].cfg1.dmaen = 1;
+  adc->adc[0].cfg1.dmacfg = 1;
+  adc->adc[0].cfg1.cont = 1;
+
+  // Enable ADC oversampling.
+  adc->adc[0].cfg2.rovse = (oversample_log2 > 0) ? 1 : 0;
+
+  // The over sample setting is the log base 2 of the number of samples
+  // less one.
+  // For example, for 16 samples we would use log2(16)-1 = 3
+  adc->adc[0].cfg2.ovsr = (oversample_log2 > 0) ? oversample_log2 - 1 : 0;
+
+  // The A/D data register can only return 16 bits, so we could get
+  // truncated data if our oversampling is higher then 16.  In this
+  // case I need to tell the hardware to downshift the sum
+  adc->adc[0].cfg2.ovss = (oversample_log2 < 4) ? 0 : (oversample_log2 - 4);
+
+  // Set sample time. I'm using 92.5 A/D clocks to sample.
+  static_assert(adc_samp_time == 92);
+  adc->adc[0].samp.smp5 = 5;
+  adc->adc[0].samp.smp6 = 5;
+  adc->adc[0].samp.smp9 = 5;
+  adc->adc[0].samp.smp15 = 5;
+
+  // Set the conversion sequence
+  adc->adc[0].seq.len = adc_channels - 1;
+  adc->adc[0].seq.sq1 = 6;
+  adc->adc[0].seq.sq2 = 9;
+  adc->adc[0].seq.sq3 = 15;
+
+  // I use DMA1 channel 1 to copy my A/D readings into my buffer
+  EnableClock(DMA1_BASE);
+  DMA_Regs *dma = DMA1_BASE;
+  int C1 = static_cast<int>(DMA_Chan::C1);
+
+  dma->channel[C1].pAddr = &adc->adc[0].data;
+  dma->channel[C1].mAddr = adc_buff;
+  dma->channel[C1].count = adc_samp_history * adc_channels;
+
+  dma->channel[C1].config.enable = 0;
+  dma->channel[C1].config.tcie = 0;
+  dma->channel[C1].config.htie = 0;
+  dma->channel[C1].config.teie = 0;
+  dma->channel[C1].config.dir =
+      static_cast<REG>(DmaChannelDir::PERIPHERAL_TO_MEM);
+  dma->channel[C1].config.circular = 1;
+  dma->channel[C1].config.perInc = 0;
+  dma->channel[C1].config.memInc = 1;
+  dma->channel[C1].config.psize = 1;
+  dma->channel[C1].config.msize = 1;
+  dma->channel[C1].config.priority = 0;
+  dma->channel[C1].config.enable = 1;
+
+  // Start the A/D converter
+  adc->adc[0].ctrl |= 4;
+}
+
+// Read the specified analog input.
+Voltage HalApi::analogRead(AnalogPin pin) {
+  int offset = [&] {
+    switch (pin) {
+    case AnalogPin::PATIENT_PRESSURE:
+      return 0;
+    case AnalogPin::INFLOW_PRESSURE_DIFF:
+      return 1;
+    case AnalogPin::OUTFLOW_PRESSURE_DIFF:
+      return 2;
+    }
+    // All cases covered above (and GCC checks this).
+    __builtin_unreachable();
+  }();
+
+  // We just run through the buffer and add up all the readings for
+  // this channel.  The DMA is still writing to this buffer in the
+  // background, but that shouldn't cause any problems because memory
+  // accesses for 16-bit values are atomic.
+  float sum = 0;
+  for (int i = 0; i < adc_samp_history; i++)
+    sum += adc_buff[i * adc_channels + offset];
```

Comment:
Because this is being written in the background, doesn't the array need to be volatile?

Reply:
Good point.  How did I miss that, I love marking things volatile!